payments-core/src/main/java/com/stripe/android/cards/DefaultCardAccountRangeRepository.kt

```diff
@@ -10,28 +10,31 @@ internal class DefaultCardAccountRangeRepository(
     private val staticSource: CardAccountRangeSource,
     private val store: CardAccountRangeStore
 ) : CardAccountRangeRepository {
+
     override suspend fun getAccountRange(
         cardNumber: CardNumber.Unvalidated
     ): AccountRange? {
-        return cardNumber.bin?.let { bin ->
-            if (store.contains(bin)) {
-                inMemorySource.getAccountRange(cardNumber)
-            } else {
-                remoteSource.getAccountRange(cardNumber)
-            } ?: staticSource.getAccountRange(cardNumber)
+        return getAccountRanges(cardNumber).orEmpty().firstOrNull {
+            it.binRange.matches(cardNumber)
         }
     }
 
     override suspend fun getAccountRanges(
         cardNumber: CardNumber.Unvalidated
     ): List<AccountRange>? {
-        return cardNumber.bin?.let { bin ->
-            if (store.contains(bin)) {
-                inMemorySource.getAccountRanges(cardNumber)
-            } else {
-                remoteSource.getAccountRanges(cardNumber)
-            } ?: staticSource.getAccountRanges(cardNumber)
+        val bin = cardNumber.bin ?: return null
+
+        val ranges = if (store.contains(bin)) {
+            inMemorySource.getAccountRanges(cardNumber)
+        } else {
+            remoteSource.getAccountRanges(cardNumber)
         }
+
+        if (!ranges.isNullOrEmpty()) {
+            store.save(bin, ranges)
+        }
+
+        return ranges?.takeIf { it.isNotEmpty() } ?: staticSource.getAccountRanges(cardNumber)
```

Comment:
We only save the response if we actually get a non-empty one. The backend service returns an empty list for some valid BINs (🤷) and I assume we don’t want to serve those but would rather fall back to the static ranges.

Reply:
I do think that the remove backend service is the source of truth, we probably should not rely on our static list. I think as long as the response is successful, we should use that data.

Reply:
The BIN data is often wrong, even in cases where it _doesn't_ return an empty response. (See RUN_MOBILESDK-2135 for more discussion on this.)

On iOS in the card field, we only fetch it for UnionPay BINs as they could be 19 digit cards (81/62). For this `possibleCardBrands` API, we actually maintain a separate cache, as we don't want to pollute the card text field's cache with bad data. Would that explain the difference in behavior you're seeing?